### PR TITLE
Lock contention on arcs mtx

### PR DESCRIPTION
**Pull request by behlendorf**
The finalized patch stack for the ARC mutex lock contention changes.

### CODE REVIEW

**Comment by FransUrbo** [comment]
In short, what does this all mean in plain english? :)

**Comment by behlendorf** [comment]
Significantly reduced lock contention in the ARC which should translate in to improved cache performance once a couple other bottlenecks are removed.  But this was a biggie.

**Comment by FransUrbo** [comment]
Ok, thanx.

**Comment by FransUrbo** [comment]
Maybe we should have cut a 0.6.5 first, before accepting this… This seems like a big thing, and it will take some time for it to 'percolate' on peoples machines before we can cut a new version.

To late now, but something to think about for next time perhaps?

**Comment by odoucet** [comment]
well, it depends on when you were expecting 0.6.5 ... there is ~ one release per year, so if 0.6.5 is released in february 2016, we will have plenty of time to check this commit :)

**Comment by FransUrbo** [comment]
> there is ~ one release per year
> 
That was just hugely bad luck. We've said we'd try to do one every three months (about one month left on that :), but this one seems to depend on others (or rather, others depend on this one) and maybe that's not enough time.

**Comment by AndCycle** [comment]
I can live with unstable release that been tested over a week without losing data :x

**Comment by behlendorf** [comment]
Right, I'd like to shoot for every 3 months or so but it looks like we might run a bit over because we'd planned to also get the ABD changes in the the next tag too.  Which is what I'm going to push to get merged that.  That way the next major tag will contain a big chunk of the memory / ARC restructuring which is a big step forward.

In the meanwhile we'll definitely keep the release branch up to date with critical fixes.  Which means right now we have a few build issues for newer kernels which need to be finalized.

**Comment by FransUrbo** [comment]
> Right, I'd like to shoot for every 3 months or so but it looks like we might run a bit over because we'd planned to also get the ABD changes in the the next tag too.
> 
Is this safe? Only having 'a few weeks' (depending on how soon you can get the ABD changes in) of testing in master before tagged as a release?

Technically (according to the schedule), we're only about four, five weeks away from next release…

**Comment by behlendorf** [comment]
@FransUrbo it's definitely ambitious and it likely means the release date will slide a month or two.  But this is one of the major items planned for the next tag.  Roughly the full remaining planned list looks like this.  We could bump ABD to the next tag but I'm inclined for the moment to try and get it finalized and merged.  If that turns out to be unworkable we can bump it.

* ABD (optional)
* Linux 4.1 build fixes (mandatory)
* Fix .zfs/snapshot for newer kernels (mandatory)
* zfs-initramfs (optional)
* Any critical bug fixes (mandatory)

As an aside what do you think about re-enabling the wiki and using it to post this kind of information.
